core/src/main/java/dk/alexandra/fresco/framework/util/DetermSecureRandom.java

```diff
@@ -1,89 +1,133 @@
+   * @throws NoSuchAlgorithmException If the <code>algorithm</code> is not found on the system.
    */
-  public DetermSecureRandom(int amount, byte[] seed) {
-    this(amount, seed, "SHA");
-  }
-
-  protected DetermSecureRandom(int amount, byte[] seed, String algorithm) {
-    try {
-      md = MessageDigest.getInstance(algorithm);
-    } catch (NoSuchAlgorithmException e) {
-      throw new MPCException("The algorithm " + algorithm + " was not found");
-    }
-    if (amount >= md.getDigestLength() || amount < 1) {
-      throw new IllegalArgumentException(
-          "Amount was set to " + amount + " but must be between 1 and " + md.getDigestLength());
+  public DetermSecureRandom(byte[] seed, String algorithm) throws NoSuchAlgorithmException {
+    this.algorithm = algorithm;
+    this.mac = Mac.getInstance(algorithm);
+    this.key = new byte[64];
+    this.val = new byte[64];
+    for (int i = 0; i < this.val.length; i++) {
+      this.val[i] = 1;
     }
-    this.amount = amount;
+    initializeMac(this.key);
     setSeed(seed);
   }
 
-  /**
-   * Convenience constructor. It does the same as a call to {@link #DetermSecureRandom(int, byte[])}
-   * with the parameters (1, new byte[]{0x00}) would do.
-   */
-  public DetermSecureRandom() {
-    this(1, new byte[] { 0x00 });
-  }
+
 
   @Override
   public synchronized void nextBytes(byte[] bytes) {
-    byte[] res = null;
-    this.md.update(this.seed);
-    res = this.md.digest();
-    int index = 0;
-    for (int i = 0; i < bytes.length; i++) {
-      bytes[i] = res[index++];
-      if (index >= amount) {
-        this.md.reset(); // ?
-        byte[] newEntropy = new byte[md.getDigestLength() - amount];
-        this.md.update(newEntropy);
-        res = md.digest();
-        index = 0;
+    int pos = 0;
+    while (pos < bytes.length) {
+      this.val = this.mac.doFinal(this.val);
+      int length = val.length;
+      //Ensure that we don't break boundries
+      if (length > bytes.length - pos) {
+        length = bytes.length - pos;
       }
+      System.arraycopy(val, 0, bytes, pos, length);
+      pos += length;
+    }
+    update(null);
+    // If we at some point want to introduce an indicator that a reseed should happen, we need to
+    // have a reseedCounter counted here. The reseed interval is 2^48 as per the NIST details.
+    // reseed_counter++;
+  }
+
+  private void update(byte[] providedData) {
+    this.mac.update(val);
+    this.key = this.mac.doFinal(new byte[] {0x00});
+    initializeMac(this.key);
+    this.val = this.mac.doFinal(this.val);
+    if (providedData == null) {
+      return;
+    }
+    this.mac.update(this.val);
+    this.mac.update(new byte[] {0x01});
+    this.key = this.mac.doFinal(providedData);
+    initializeMac(this.key);
+    this.val = mac.doFinal(this.val);
+  }
+
+  private void initializeMac(byte[] key) {
+    try {
+      this.mac.init(new SecretKeySpec(key, this.algorithm));
```

Comment:
Create a package local method called createKey that takes a Supplier of SecretKeySpec and catches the exception. Then test it with a different supplier.

Reply:
Oh well - the idea is here, the overall comment is ... overall

Reply:
Yea okay. That is a solution, but I don't ever see a point in having more than one supplier - other than for test purposes. So still feels kinda hacky.

Reply:
Feelings are normally dealt with at the psychologist or a coach :)

A supplier is a functional interface - it is just a lambda. A test method would look like
createKey(()->throw new InvalidKeyException())
Or even
createKey(()->new SecretKey(new byte[0]))

I think it is a perfectly valid solution, however there are definitely room for flexibility within it. Feel free to suggest something.

I **feel** it is pretty lame not to test the exception case.  :)